scripts/operations/configuration/update_product_catalog_ims_ids.py

```diff
@@ -0,0 +1,156 @@
+    if parsed_args.ims_type == 'image':
+        return old_new_map, {}
+    if parsed_args.ims_type == 'recipe':
+        return {}, old_new_map
+    # This should never happen, but just in case
+    raise UpdateProductCatalogImsIdsBaseError(
+        "PROGRAMMING LOGIC ERROR: Unable to determine type of IMS ID to replace")
+
+def get_name_id_replacements(product_name: str, product_version: str,
+                             product_version_details: JSONObject, images_or_recipes: str,
+                             id_change_map: IdChangeMap) -> product_catalog.NameIdTupleList:
+    """
+    Looks at the specified product version entry from the Cray Product Catalog.
+    Based on the 'images_or_recipes' parameter, looks at the 'images' or 'recipes' field of the
+    entry. If no such field exists, then there are no IDs to be updated. If the field exists,
+    its contents are scanned for IDs that match ones in the specified id_change_map.
+    Returns a list of tuples: (image_or_recipe_name, new_ims_id_for_this_image_or_recipe)
+    """
+    name_id_list = []
+    if not id_change_map:
+        # Nothing to do
+        return name_id_list
+    if not images_or_recipes in product_version_details:
+        # No field for images or recipes
+        return name_id_list
+    prod_ver_imgrec_map = product_version_details[images_or_recipes]
+
+    # The contents of this field should be a dictionary
+    if not isinstance(prod_ver_imgrec_map, dict):
+        LOGGER.debug("Product catalog entry for '%s' version '%s' contains '%s' field which is "
+                     "type %s instead of dict", product_name, product_version, images_or_recipes,
+                     str(type(prod_ver_imgrec_map)))
+        return name_id_list
+
+    # This dictionary should map image/recipe names to another dictionary, and that
+    # second dictionary should include an id field, which maps to the IMS ID for that image/recipe
+    # So parse through this and look for any IDs that need to be replaced.
+
+    for imgrec_name, imgrec_data in prod_ver_imgrec_map.items():
+        if not isinstance(imgrec_data, dict):
+            LOGGER.debug("Product catalog entry for '%s' version '%s' contains %s '%s' with "
+                         "unexpected data format: %s", product_name, product_version,
+                         images_or_recipes, imgrec_name, str(imgrec_data))
+            continue
+        if "id" not in imgrec_data:
+            LOGGER.debug("Product catalog entry for '%s' version '%s' contains %s '%s' with no "
+                         "'id' data: %s", product_name, product_version, images_or_recipes,
+                         imgrec_name, str(imgrec_data))
+            continue
+        imgrec_id = imgrec_data["id"]
+        if imgrec_id in id_change_map:
+            # This means that this ID was changed during the import
+            name_id_list.append((imgrec_name, id_change_map[imgrec_id]))
```

Comment:
Generally speaking, I don't think this should happen given that an IMS export/import is going to change the IMS IDs of all recipes and images that were recorded in the product catalog. However, if there were any entries for images or recipes whoser IDs did not change, it looks like they'd be omitted from the `name_id_list` return value here. They would then be left out of the `YAML_CONTENT_STRING` passed to the `cray-product-catalog-update` container. I don't remember if `cray-product-catalog-update` merges the dictionaries under the `images` and `recipes` keys. I think older versions certainly did not merge and just replaced the values completely, but newer versions of `cray-product-catalog-update` may handle this better now. It might be worth double-checking, just to be sure we aren't going to drop information from the cray-product-catalog.

Reply:
For the 1.3 backport I made a change so that it does the merging on my end, thus making it unimportant whether or not the updater handles it.